grpc-for-smart-contracts.md

```diff
@@ -0,0 +1,263 @@
+A transaction signature is map from the index of the credential to another map from the key index to the actual signature.
+The credential index is relative to the account address, and the indices should be distinct.
+The key index is relative to the credential.
+The maximum length of the list is 255, and the minimum length is 1.
+`TransactionSignature = Map CredentialIndex (Map KeyIndex Signature)`
+
+#### Serialization
+- Length of outer map (uint8)
+- For each keypair in outer map:
+  - `CredentialIndex` (uint8)
+  - Length of inner map (uint8)
+  - For each keypair in inner map:
+    - `KeyIndex` (uint8)
+    - Length of `Signature` in bytes (uint16, big-endian)
+    - `Signature` as bytes
+
+### `TransactionHeader`
+
+A transaction header is a struct which consists of the following fields:
+
+`AccountAddress`: The sender account.
+
+`Nonce`: Account nonce. Is incremented by 1 with every transaction originating
+from an account.
+
+`Energy`: The amount of energy allocated for the execution of this transaction.
+
+`PayloadSize`: Size of the `Payload` in bytes.
+
+`TransactionExpiryTime`: Absolute expiration time after which transaction will
+*not* be executed. Measured in seconds since unix epoch.
+
+#### Serialization
+
+- `AccountAddress` (byte array of size 32)
+- `Nonce` (uint64, big-endian)
+- `Energy` (uint64, big-endian)
+- `PayloadSize` (uint32, big-endian)
+- `TransactionExpiryTime` (uint64, big-endian)
+
+### `Payload`
+A `Payload` is a union type.
+The three relevant variants for smart contracts are `DeployModule`,
+`InitContract`, and `Update`.
+
+#### Serialization
+
+- Tag (uint8)
+  - Relevant tag values:
+    - `0`: `DeployModule`
+    - `1`: `InitContract`
+    - `2`: `Update`
+- Content of variant (see below)
+
+### DeployModule
+
+`WasmModule` contains a version number and the binary module source.
+
+#### Serialization
+
+ - Version (uint32, big-endian)
+ - Length of byte array (uint32, big-endian)
+ - The byte array
+
+### InitContract
+
+`Amount`: Amount in microGTU (10^-6 GTU).
+
+`ModuleRef`: Hash of the module on chain.
+
+`InitName`: Name of init function including `init_` prefix.
+
+`Parameter`: Parameter for the init function.
+
+#### Serialization
+
+- `Amount` (uint64, big-endian)
+- `ModuleRef` (byte array of fixed size 32)
+- `InitName`
+  - Length of name in bytes (uint16, big-endian)
+  - Name as a byte array
+- `Parameter` (depends on the contract)
+
+### Update (Contract)
+`Amount`: Amount in microGTU (10^-6 GTU).
+
+`ContractAddress`: Address of contract instance consisting of an index and a subindex.
+
+`ReceiveName`: Name of receive function including `<contractName>.` prefix.
+
+`Parameter`: Parameter for the receive function.
+
+#### Serialization
+
+- `Amount` (uint64, big-endian)
+- `ContractAddress`
+  - Index (uint64, big-endian)
+  - Subindex (uint64, big-endian)
+- `ReceiveName`
+  - Length of name in bytes (uint64, big-endian)
+  - Name as byte array
+- `Parameter` (depends on the contract)
```

Comment:
Parameter serialization is always fixed, it's only the meaning of the parameter that is dependent on the  contract.